keps/sig-network/4427-relaxed-dns-search-validation/README.md

```diff
@@ -0,0 +1,484 @@
+that it's better to stick with valid host names but also states that labels must meet the hostname requirements.
+
+In practice, legacy workloads sometimes include an underscore (`_`) in DNS names and DNS servers will generally allow this.
+
+This KEP proposes relaxing the checks on DNS search strings only. Allowing these values in the `searches` field of `dnsConfig` allows pods to
+resolve short names properly in cases where the search string contains an underscore.
+
+## Motivation
+
+For workloads that resolve short DNS names where the full DNS name includes underscores,
+it’s not possible to configure search strings using dnsConfig. For example, if a pod needs to look up an SRV record `_sip._tcp.abc_d.example.com`
+using the short name of `_sip._tcp`, we would like to be able to add `abc_d.example.com` to the searches in the dnsConfig.
+
+Here’s an example configuration which would support this case:
+
+```
+apiVersion: v1
+kind: Pod
+metadata:
+  namespace: default
+  name: dns-example
+spec:
+  containers:
+    - name: test
+      image: nginx
+  dnsPolicy: "None"
+  dnsConfig:
+    nameservers:
+      - 1.2.3.4
+    searches:
+      - abc_d.example.com
+```
+
+However, this returns an error:
+
+```
+The Pod "dns-example" is invalid: spec.dnsConfig.searches[0]: Invalid value: "abc_d.example.com": a DNS-1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*')
+```
+
+Allowing underscores in the search string allows integration with legacy workloads without allowing anyone to define
+these names within Kubernetes. Since having underscores in a name creates other issues (such as inability to obtain a publicly trusted TLS certificate),
+search strings seem like the only area where this is likely to occur.
+
+### Goals
+
+- Support workloads that need to resolve DNS short names where the full DNS name includes an underscore (`_`).
+
+## Proposal
+
+Introduce a RelaxedDNSSearchValidation feature gate which is disabled by default. When the feature gate is enabled,
+a new DNS name validation function will be used, which keeps the existing validation logic but also allows an underscore (`_`) in any place
+where a dash (`-`) would be allowed currently.
+
+Since the relaxed check allows previously invalid values, care must be taken to support cluster downgrades safely. To accomplish this, the validation will distinguish between new resources and updates to existing resources:
+- When the feature gate is disabled:
+  - (a) New resources will use strict validation based on RFC-1123 (no change to current validation)
+  - (b) Updates to existing resources will use relaxed validation if any search string in the existing list fails strict validation
```

Comment:
@thockin @danwinship you can not update the pod.spec.dnsConfig

Reply:
ok, but you can do it in the deployments as in `podTemplate` and that will trigger a pod restart

Reply:
yes, in PodTemplates.

Reply:
hmm, no we can not do this,  because the pods will fail to be created

Reply:
What do you mean?  Updating a deployment will make a new RS and all new pods.  There's no in-place update of pods happening.

Reply:
Oh - that's an interesting one. So the validation on the pod can't stand on it's own, since the existing value may be from a pod template.

Reply:
assume I have a deployment with 
```
  template:
    metadata:
      labels:
        app: MyApp
    spec:
      containers:
      - args:
        - netexec
        - --http-port=7070
        - --udp-port=7070
        image: k8s.gcr.io/e2e-test-images/agnhost:2.41
        imagePullPolicy: IfNotPresent
        name: agnhost
        ports:
        - containerPort: 7070
          protocol: TCP
      dnsPolicy: ClusterFirst
        dnsConfig:
    nameservers:
    - 1.2.3.4
    searches:
    - abc-d.example.com
```

so, any update to this `podTemplate` that is not going to change the `searches` to a value that matches `strict` validation, will trigger the creation of new Pods using a `search` string that will fail, because new resources wil use `strict` validation

Reply:
Ahh, I see what you mean.  Hmm.  I don't know what we have done here in the past.  Specifically:

Forget updates, it's simpler:

I have a Deployment which uses "a_b.example.com".  I disable the gate.  I do a rollout restart.  Each new pod will fail CREATE because the gate disallows "_"

@liggitt O keeper of memories, do you recall any such similar cases?

Reply:
I don't think we've done anything special for parent → child create cases where the create gets rejected, we rely on controller backoff or surfacing the inability to make progress back up to the parent object.

Reply:
@thockin what are your thoughts on handling this? This would increase the risk on downgrades (negating some of the ratcheting logic, since pods are generally created via a template). It would still protect against the scenario where you have an object that is already saved but can't be updated.

Approaches that come to mind:
- Always relax the check on pods but keep the ratcheting on podTemplates. This allows underscores outside the feature gate on pods that are created directly.
- Proceed as proposed, but document the risk identified here.
- Special case validation when a pod is created from a template (not sure this is really possible).